### ISSUE TITLE
[feature request] AttentionCellWrapper

### ISSUE THREAD

**Issue by jendrikjoe** [pytorch/pytorch#6294]
Dear PyTorch Community,

I am currently implementing RNNs with an attention mechanism out of interest.
Since I wanted to implement it in pyTorch I came up with the following class to wrap RNNCells:

    class AttentionCellWrapper(Module):
        r"""Create a cell with attention to past data samples.
        Args:
    
            cell: an RNNCell, an attention is added to it.
            attnWindowSize: integer, the size of an attention window.
            mlpLayerSize: integer, the size of the neural network
            to determine the attention weights.
    
        Attributes:
            weightList: the weighting of the inputs from the last forward pass
    
        Examples::
            encoder = LSTMCell(inputSize, 64)
            decoderCell = LSTMCell(hiddenSize, encoder.hidden_size)
            decoder = AttentionCellWrapper(decoderCell,
                                    12, 128)
            outputs = []
            h_t = Variable(torch.zeros(x.size(1),
                                       encoder.hidden_size).float(),
                           requires_grad=False).cuda()
            c_t = Variable(torch.zeros(x.size(1),
                                       encoder.hidden_size).float(),
                           requires_grad=False).cuda()
            for i in range(x.size(0)):
                h_t, c_t = encoder(x[i], (h_t, c_t))
                outputs += [h_t.clone()]
            outputs = torch.stack(outputs, 0).squeeze(2)
            outputs = decoder(outputs)
            weightList = decoder.weightList
        """
    
        def __init__(self, cell, attnWindowSize, mlpLayerSize=100):
            super(AttentionCellWrapper, self).__init__()
            self.cell = cell
            self.attnLength = cell.input_size
            self.attnMlp = Linear(self.cell.hidden_size, mlpLayerSize)
            self.weightMlp = Linear(2*mlpLayerSize, 1)
            self.attnWindowSize = attnWindowSize
            self.reset_parameters()
            self.weightList = []
    
        def reset_parameters(self):
            self.attnMlp.reset_parameters()
            self.weightMlp.reset_parameters()
    
        def forward(self, inputVec):
            hT = Variable(
                torch.zeros(inputVec.size(1), self.cell.hidden_size).float(),
                requires_grad=False).cuda()
            cT = Variable(
                torch.zeros(inputVec.size(1), self.cell.hidden_size).float(),
                requires_grad=False).cuda()
            outputs = []
            weightList = []
            # Parameterize Input
            inputVec = self.attnMlp(inputVec)
            for i in range(inputVec.size(0)):
                if i == 0:
                    hT, cT = self.cell(inputVec[i], (hT, cT))
                    outputs += [hT.clone()]
                else:
                    # Parameterize LSTM State
                    attnVec = self.attnMlp(hT)
                    # Stack all previous inputs
                    if i-self.attnWindowSize >= 0:
                        stackedInputs = inputVec[i-self.attnWindowSize+1:i+1]
                    else:
                        stackedInputs = inputVec[:i+1]
                    # Calculate the weights for the previous inputs
                    attnVec = attnVec.unsqueeze(0).expand(*stackedInputs.size())
                    catenated = torch.cat((attnVec, stackedInputs), dim=2)
                    weights = ELU()(self.weightMlp(catenated))
                    weights = Softmax(dim=0)(weights)
                    weightedInput = torch.sum(weights*stackedInputs, dim=0)
                    # Run LSTM step
                    hT, cT = self.cell(weightedInput, (hT, cT))
                    outputs += [hT.clone()]
                    if i - self.attnWindowSize >= 0:
                        # Store weights, if more than attnWindowSize 
                        # timesteps are in the input
                        weightList.extend(weights.clone().permute(1, 0, 2).cpu().data.numpy())
            self.weightList = weightList
            return torch.stack(outputs, dim=0)

The class currently only supports attention on past features, so it is probably more interesting for people dealing with time series, but I think it could be polished to be used for NLPs as well.
Furthermore, the alignment model is currently based on a linear network, which maybe should be changeable to other options. 
However, is this generally something that I should add using a PR request?
If so, are there any changes I should do before-hand (I assume yes ;) )?

Cheers and thanks for any feedback,

Jendrik
